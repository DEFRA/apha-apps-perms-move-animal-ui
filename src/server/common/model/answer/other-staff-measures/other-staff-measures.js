import { TextAreaAnswer } from '../text-area/text-area.js'

/** @import {TextAreaConfig} from '../text-area/text-area.js' */

/**
 * @typedef {{ otherStaffMeasures: string }} OtherStaffMeasuresPayload
 */

/**
 * @augments {TextAreaAnswer<OtherStaffMeasuresPayload>}
 */
export class OtherStaffMeasuresAnswer extends TextAreaAnswer {
  /** @type {TextAreaConfig} */
  static config = {
    payloadKey: 'otherStaffMeasures',
    rows: 6,
    validation: {
      maxLength: {
        value: 5000,
        message: 'Your answer must be no longer than 5000 characters'
      },
      empty: {
        message:
          'Enter what other measures staff are taking to reduce the risk of spreading TB'
      }
    }
  }
}
