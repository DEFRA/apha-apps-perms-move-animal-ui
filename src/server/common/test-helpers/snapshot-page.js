import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
