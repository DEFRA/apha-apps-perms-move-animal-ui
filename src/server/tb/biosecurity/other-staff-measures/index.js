import { QuestionPage } from '../../../common/model/page/question-page-model.js'
import { QuestionPageController } from '../../../common/controller/question-page-controller/question-page-controller.js'
import { OtherStaffMeasuresAnswer } from '../../../common/model/answer/other-staff-measures/other-staff-measures.js'
import { badgersPage } from '../badgers/index.js'

export class OtherStaffMeasuresPage extends QuestionPage {
  urlPath = '/biosecurity/other-staff-measures'
  sectionKey = 'biosecurity'
  question =
    'What other measures are staff taking to reduce the risk of spreading TB?'

  questionKey = 'otherStaffMeasures'

  Answer = OtherStaffMeasuresAnswer

  nextPage() {
    return badgersPage
  }
}

export const otherStaffMeasuresPage = new OtherStaffMeasuresPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const otherStaffMeasures = new QuestionPageController(
  otherStaffMeasuresPage
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
