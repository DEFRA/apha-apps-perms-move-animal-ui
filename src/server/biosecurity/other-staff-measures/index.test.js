import { OtherStaffMeasuresAnswer } from '../../common/model/answer/other-staff-measures/other-staff-measures.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { badgersPage } from '../badgers/index.js'
import { otherStaffMeasuresPage, OtherStaffMeasuresPage } from './index.js'

const sectionKey = 'biosecurity'
const question =
  'What other measures are staff taking to reduce the risk of spreading TB?'
const questionKey = 'otherStaffMeasures'
const pageUrl = '/biosecurity/other-staff-measures'

describe('OtherStaffMeasures', () => {
  const page = new OtherStaffMeasuresPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(OtherStaffMeasuresAnswer)
  })

  it('nextPage should return badgers page', () => {
    const nextPage = page.nextPage()
    expect(nextPage).toBe(badgersPage)
  })

  it('should export page', () => {
    expect(otherStaffMeasuresPage).toBeInstanceOf(OtherStaffMeasuresPage)
  })

  describePageSnapshot({
    describes: 'otherStaffMeasures.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
