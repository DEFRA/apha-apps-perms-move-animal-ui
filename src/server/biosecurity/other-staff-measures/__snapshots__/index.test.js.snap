// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtherStaffMeasures otherStaffMeasures.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="otherStaffMeasures">
      What other measures are in place to clean and disinfect staff to reduce the risk of spreading TB?
    </label>
  </h1>
  <textarea class="govuk-textarea" id="otherStaffMeasures" name="otherStaffMeasures" rows="6"></textarea>
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
